### ISSUE TITLE
sql,jobs: slow quiesce on server shutdown

### ISSUE THREAD

**Issue by adityamaru** [cockroachdb/cockroach#99666]
Recently, we have seen an increase in tests timing out because of a `slow quiesce` on server shutdown. A lot of the time the stack included calls in `pkg/jobs` that resulted in #99609. However, that diff too does not seem to fix the root cause.  In https://teamcity.cockroachdb.com/viewLog.html?buildId=9272246&tab=buildResultsDiv&buildTypeId=Cockroach_UnitTests_BazelUnitTests we see a schema change GC job stuck trying to close its `rowsIterator`:

```
* goroutine 146096 [semacquire, 2 minutes]:
* sync.runtime_Semacquire(0xc008b9fe30?)
* 	GOROOT/src/runtime/sema.go:62 +0x25
* sync.(*WaitGroup).Wait(0x34d2be0?)
* 	GOROOT/src/sync/waitgroup.go:139 +0x52
* github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Close.func1()
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:487 +0x30
* github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Close(0xc0031dd0e0)
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:496 +0xde
* github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Next.func1()
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:406 +0x35
* github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Next(0x50d0443?, {0x64d4f00?, 0xc00684dbc0?})
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:416 +0x40b
* github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).queryInternalBuffered(0x0?, {0x64d4f00, 0xc00684dbc0}, {0x50f15e8, 0xe}, 0x0?, {{{0x50d0443, 0x4}}, {0x0, 0x0}, ...}, ...)
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:570 +0x1ca
* github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowExWithCols(0xc0063f9d00?, {0x64d4f00?, 0xc00684dbc0?}, {0x50f15e8?, 0x436b45?}, 0x8?, {{{0x50d0443, 0x4}}, {0x0, 0x0}, ...}, ...)
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:622 +0xc5
* github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowEx(0x45edb0?, {0x64d4f00?, 0xc00684dbc0?}, {0x50f15e8?, 0xc008ba03f8?}, 0x4431f3?, {{{0x50d0443, 0x4}}, {0x0, 0x0}, ...}, ...)
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:608 +0xb2
* github.com/cockroachdb/cockroach/pkg/jobs.Updater.load({0xc005abeb60, {0x6526d80, 0xc007f765a0}}, {0x64d4f00, 0xc00efb3200})
* 	github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:865 +0x54c
* github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).LoadJobWithTxn(0xc00324cd80, {0x64d4f00, 0xc00efb3200}, 0xbd117b6e2e48001, {0x6526d80, 0xc007f765a0})
* 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:880 +0x8d
* github.com/cockroachdb/cockroach/pkg/sql/gcjob.initDetailsAndProgress.func1({0x64d4f00?, 0xc00efb3200?}, {0x6526d80?, 0xc007f765a0?})
* 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job_utils.go:73 +0x8a
* github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn.func1({0x64d4f00?, 0xc00efb3200?}, 0x65362e0?)
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1422 +0x31
* github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4({0x64d4f00, 0xc00efb3200}, 0xc005a55340)
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1500 +0x3b4
* github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1({0x64d4f00?, 0xc00efb3200?}, 0xc009d53f20?)
* 	github.com/cockroachdb/cockroach/pkg/kv/db.go:965 +0x27
* github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc005a55340, {0x64d4f00, 0xc00efb3200}, 0xc001ef4ce8)
* 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:928 +0xae
* github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x64d4f00, 0xc00efb3200}, 0xc00efb3200?, 0x0?)
* 	github.com/cockroachdb/cockroach/pkg/kv/db.go:964 +0x6b
* github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl(0x7f6ab3ea4878?, {0x64d4f00, 0xc00efb3200}, 0x8538ac0?, 0xc0?, 0x0?, 0x0?)
* 	github.com/cockroachdb/cockroach/pkg/kv/db.go:927 +0xa7
* github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0x439247?, {0x64d4f00?, 0xc00efb3200?}, 0x1?)
* 	github.com/cockroachdb/cockroach/pkg/kv/db.go:902 +0x2d
* github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn(0xc007e37560, {0x64d4f00, 0xc00efb3200}, 0xc0063f9ce0, {0x0, 0x0, 0x4e109e0?})
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1488 +0x2a7
* github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn(0x580?, {0x64d4f00, 0xc00efb3200}, 0xc00684da40, {0x0, 0x0, 0x0})
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1423 +0x9f
* github.com/cockroachdb/cockroach/pkg/sql/gcjob.initDetailsAndProgress({0x64d4f00, 0xc00efb3200}, 0xc005959700, 0xbd117b6e2e48001)
* 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job_utils.go:71 +0x11a
* github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.Resume({0xc0032169c0}, {0x64d4f00, 0xc00efb3200}, {0x4e5c4e0?, 0xc00accbd00})
* 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:302 +0x248
* github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2(0xc0032169c0?, 0x0?, {0x64ad4e0?, 0xc00accbce0?}, {0x64d4f00?, 0xc00efb3200?}, {0x4e5c4e0?, 0xc00accbd00?}, 0xc001ef5860)
* 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1534 +0xcc
* github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc00324cd80, {0x64d4f00, 0xc00a90f560}, {0x4e5c4e0, 0xc00accbd00}, {0x64ad4e0, 0xc00accbce0}, 0xc0032169c0, {0xc0072c6c98, 0x7}, ...)
* 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1535 +0x8ed
* github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc00324cd80, {0x64d4e58, 0xc00695aa00}, {0x64ad4e0, 0xc00accbce0}, 0xc0032169c0, {0xc0072c6c98, 0x7}, {0xc00b2f8330, 0x16})
* 	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:470 +0x5d7
* github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func2({0x64d4f00?, 0xc00a90f2f0?})
* 	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:388 +0x146
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
* created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
* 
```

This issue tracks the investigation into what is preventing graceful server shutdown.

Jira issue: CRDB-26009

**Comment by yuzefovich** [comment]
Fix to master has been merged as part of #99711, will close this issue once 23.1 backport is merged.

**Comment by rafiss** [comment]
Is this different from https://github.com/cockroachdb/cockroach/issues/99242 ?

**Comment by yuzefovich** [comment]
It is different from #99242 - this issue is about a deadlock in `execinfra.RowChannel` (at least it is in my mind about that). Perhaps the issue description here is confusing and I should've kept #98837 as the tracking issue for the problem fixed in the second commit of #99609.

**Comment by yuzefovich** [comment]
Backport of the commit that fixes this was the second commit in #99970.
